src/leader.c

```diff
@@ -392,9 +332,129 @@ static int leaderApplyFrames(struct exec *req,
 	return rv;
 }
 
-static void leaderExecV2(struct exec *req, enum pool_half half)
+enum {
+	BARRIER_START,
+	BARRIER_PASSED,
+	BARRIER_DONE,
+	BARRIER_FAIL,
+	BARRIER_NR,
+};
+
+#define A(ident) BITS(BARRIER_##ident)
+#define S(ident, allowed_, flags_) \
+	[BARRIER_##ident] = { .name = #ident, .allowed = (allowed_), .flags = (flags_) }
+
+static const struct sm_conf barrier_states[BARRIER_NR] = {
+	S(START,  A(PASSED)|A(DONE)|A(FAIL), SM_INITIAL),
+	S(PASSED, A(DONE)|A(FAIL),           0),
+	S(DONE,   0,                         SM_FINAL),
+	S(FAIL,   0,                         SM_FINAL|SM_FAILURE),
+};
+
+#undef S
+#undef A
+
+static void barrier_done(struct barrier *barrier, int status)
+{
+	PRE(barrier != NULL);
+	int state = sm_state(&barrier->sm);
+	PRE(state == BARRIER_START || state == BARRIER_PASSED);
+	void (*cb)(struct barrier *, int) = barrier->cb;
+	PRE(cb != NULL);
+
+	if (status != 0) {
+		sm_fail(&barrier->sm, BARRIER_FAIL, status);
+	} else {
+		sm_move(&barrier->sm, BARRIER_DONE);
+	}
+	sm_fini(&barrier->sm);
+	/* TODO(cole) uncommment this once the barrier-callback-runs-twice
+	 * issue is fixed. */
+	/* barrier->req.data = NULL; */
+	barrier->leader = NULL;
+	barrier->cb = NULL;
+
+	if (state == BARRIER_PASSED) {
+		cb(barrier, status);
+	}
+}
+
+static void barrier_raft_cb(struct raft_barrier *, int);
+
+static int barrier_async(struct barrier *barrier, int status)
```

Comment:
Nit: Isn't this more like a `barrier_tick`? It moves the state machine and is called when needed both from aync and sync contexts.

Reply:
Yeah, `tick` is a good name.